src/libbson/src/bson/bson-error.c

```diff
@@ -109,12 +109,17 @@ bson_strerror_r (int err_code,  /* IN */
    if (strerror_s (buf, buflen, err_code) != 0) {
       ret = buf;
    }
-#elif defined(__GNUC__) && defined(_GNU_SOURCE)
-   ret = strerror_r (err_code, buf, buflen);
-#else /* XSI strerror_r */
+#elif defined(_XOPEN_SOURCE) && _XOPEN_SOURCE >= 600 && !defined(_GNU_SOURCE)
```

Comment:
> What is the argument for not using the logic suggested in [strerror_r(3)](https://linux.die.net/man/3/strerror_r)?

`_XOPEN_SOURCE` implies `_POSIX_C_SOURCE` per https://pubs.opengroup.org/onlinepubs/7908799/xsh/compilation.html: "Since this specification is aligned with the ISO C standard, and since all functionality enabled by _POSIX_C_SOURCE set greater than zero and less than or equal to 199506L should be enabled by _XOPEN_SOURCE set equal to 500, there should be no need to define either _POSIX_SOURCE or _POSIX_C_SOURCE if _XOPEN_SOURCE is so defined."

> Likewise, `!_GNU_SOURCE` would also handle the case where `_GNU_SOURCE` is defined but zero (not sure how relevant that is).

Only the presence of `_GNU_SOURCE` matters, not its value, per https://man7.org/linux/man-pages/man7/feature_test_macros.7.html: "Defining this macro (with any value) implicitly defines [...]. In addition, various GNU-specific extensions are also exposed."

> I don't see `_POSIX_SOURCE` referred to anywhere within libmongoc.

It is unconditionally defined via the definition of `_XOPEN_SOURCE` here: https://github.com/mongodb/mongo-c-driver/blob/6d2be25206daeb55960c84a5f129b226f0a25393/CMakeLists.txt#L398

All our requirements concerning POSIX and extensions are being handled by the (unconditional) definition of the macros listed in the link above. The conflict between the XSI-compliant version and GNU extension version of `strerror_r` is AFAIK a unique case.

Reply:
Thanks for the rigorous response! LGTM.